### PR TITLE
Xray Cloud Integration

### PR DESCRIPTION
**Pull request by amitasil**

### CODE REVIEW

**Comment by amitasil** [comment]
@medtoure18 

Please review this PR once you get some time.

I am able to generate the required Access Token from the first request and passing it to the second API request but I am not able to change the Test Execution status in XRay.
However, if I use the same request which I am sending through this code and paste it in Postman, I am able to change the status of Test Execution in XRay.

Body for the second request (working in Postman but not through the code):
`{ "testExecutionKey": "SP-10","info" : {"startDate" : "2021-11-24T21:31:06+11:00", "finishDate" :"2021-11-24T21:31:06+11:00","revision": "001","description" : "Results of test execution ", "testEnvironments": ["Sandpit"]},"tests" : [{"testKey":"SP-8","status":"PASSED", "examples":[],"comment" : "Successful execution" }]}`

Not sure exactly what am I missing. Maybe need to store/handle some cookies not sure.

**FYI:** you might need to increase the timeout value.

**Comment by medtoure18** [comment]
@amitasil 
there is an error in the json, it's PASS and not PASSED

`{ "testExecutionKey": "SP-10","info" : {"startDate" : "2021-11-24T21:31:06+11:00", "finishDate" :"2021-11-24T21:31:06+11:00","revision": "001","description" : "Results of test execution ", "testEnvironments": ["Sandpit"]},"tests" : [{"testKey":"SP-8","status":"PASS", "examples":[],"comment" : "Successful execution" }]}`

see xray doc
image

https://docs.getxray.app/display/XRAY/Import+Execution+Results+-+REST#ImportExecutionResultsREST-CucumberJSONresultsMultipart

File: index.js [hunk 1]
```diff
@@ -34,13 +39,13 @@ module.exports = function (config) {
                 if (executedTest.tags[1].split("@")[1] == test.tags[1].split("@")[1]) {
                     //we add the first test result status in exemples array
                     if (executedTest.examples.length == 0) {
-                        if (executedTest.state == "passed") executedTest.examples.push("PASS");
-                        else executedTest.examples.push("FAIL");
+                        if (executedTest.state == "passed") executedTest.examples.push("PASSED");
```
Comment (medtoure18) [inline index.js]: you should leaved it to pass. unless it's dont work on cloud.
in that case, you can asomething like
`if (xray_cloud) executedTest.examples.push("PASSED");
else executedTest.examples.push("PASS");`

File: index.js [hunk 2]
```diff
@@ -34,13 +39,13 @@ module.exports = function (config) {
...
+                        else executedTest.examples.push("FAILED");
```
Comment (medtoure18) [inline index.js]: same suggestion as the PASS

**Comment by amitasil** [comment]
@medtoure18 

This is the sample for xray cloud:

https://docs.getxray.app/display/XRAYCLOUD/Import+Execution+Results+-+REST+v2

image

File: index.js [hunk 3]
```diff
@@ -92,8 +97,8 @@ module.exports = function (config) {
                 if (i == 0) {
                     tests = '"tests" : [';
                     // if this is the first test that is added to the list of tests executed.
-                    tests = tests + "" + '{"testKey":"' + test.tags[1].split("@")[1] + '","status":"' + test.state + '", "examples":' + JSON.stringify(test.examples) + ',"comment" : "' + test.comment + '" }';
-                } else tests = tests + "" + ',{"testKey":"' + test.tags[1].split("@")[1] + '","status":"' + test.state + '", "examples":' + JSON.stringify(test.examples) + ',"comment" : "' + test.comment + '" }';
+                    tests = tests + "" + '{"testKey":"' + test.tags[1].split("@TEST_")[1] + '","status":"' + test.state + '", "examples":' + JSON.stringify(test.examples) + ',"comment" : "' + test.comment + '" }';
```
Comment (medtoure18) [inline index.js]: this is not ok of server version.
the tag, can be '@ZZZ-1122' or something else.
If it's always like '@TEST_'  on cloud version, you can also add an if cloud

File: index.js [hunk 4]
```diff
@@ -34,13 +39,13 @@ module.exports = function (config) {
...
                     //we add the first test result status in exemples array
                     if (executedTest.examples.length == 0) {
-                        if (executedTest.state == "passed") executedTest.examples.push("PASS");
-                        else executedTest.examples.push("FAIL");
+                        if (executedTest.state == "passed") executedTest.examples.push("PASSED");
+                        else executedTest.examples.push("FAILED");
                     }
                     //we add the actual test result status in exemples array
-                    if (test.state == "passed") executedTest.examples.push("PASS")
+                    if (test.state == "passed") executedTest.examples.push("PASSED")
```
Comment (medtoure18) [inline index.js]: same suggestion , it should be PASS or Fail, or you can add an "if cloud"

File: index.js [hunk 5]
```diff
@@ -34,13 +39,13 @@ module.exports = function (config) {
...
                     else {
-                        executedTest.examples.push("FAIL");
+                        executedTest.examples.push("FAILED");
```
Comment (medtoure18) [inline index.js]: same suggestion , it should be PASS or Fail, or you can add an "if cloud"

File: index.js [hunk 6]
```diff
@@ -49,13 +54,13 @@ module.exports = function (config) {
                 if (executedTest.tags[2].split("@")[1] == test.tags[2].split("@")[1]) {
                     //we add the first test result status in exemples array
                     if (executedTest.examples.length == 0) {
-                        if (executedTest.state == "passed") executedTest.examples.push("PASS")
-                        else executedTest.examples.push("FAIL")
+                        if (executedTest.state == "passed") executedTest.examples.push("PASSED")
```
Comment (medtoure18) [inline index.js]: same suggestion , it should be PASS or Fail, or you can add an "if cloud"

File: index.js [hunk 7]
```diff
@@ -49,13 +54,13 @@ module.exports = function (config) {
...
+                        else executedTest.examples.push("FAILED")
                     }
                     //we add the actual test result status in exemples array
-                    if (test.state == "passed") executedTest.examples.push("PASS")
+                    if (test.state == "passed") executedTest.examples.push("PASSED")
```
Comment (medtoure18) [inline index.js]: same suggestion , it should be PASS or Fail, or you can add an "if cloud"

File: index.js [hunk 8]
```diff
@@ -49,13 +54,13 @@ module.exports = function (config) {
...
                     else {
-                        executedTest.examples.push("FAIL");
+                        executedTest.examples.push("FAILED");
```
Comment (medtoure18) [inline index.js]: same suggestion , it should be PASS or Fail, or you can add an "if cloud"

File: index.js [hunk 9]
```diff
@@ -74,10 +79,10 @@ module.exports = function (config) {
         //we change test state syntax to meet xray requirement
         testsResults.forEach(test => {
             if (test.state == "passed") {
-                test.state = 'PASS';
+                test.state = 'PASSED';
```
Comment (medtoure18) [inline index.js:96]: same suggestion , it should be PASS or Fail, or you can add an "if cloud"

File: index.js [hunk 10]
```diff
@@ -74,10 +79,10 @@ module.exports = function (config) {
...
                 test.comment = "Successful execution";
             } else {// if the test is failed, we get the reason for the failure
-                test.state = 'FAIL';
+                test.state = 'FAILED';
```
Comment (medtoure18) [inline index.js:99]: same suggestion , it should be PASS or Fail, or you can add an "if cloud"

File: index.js [hunk 11]
```diff
@@ -92,8 +97,8 @@ module.exports = function (config) {
                 if (i == 0) {
                     tests = '"tests" : [';
                     // if this is the first test that is added to the list of tests executed.
-                    tests = tests + "" + '{"testKey":"' + test.tags[1].split("@")[1] + '","status":"' + test.state + '", "examples":' + JSON.stringify(test.examples) + ',"comment" : "' + test.comment + '" }';
-                } else tests = tests + "" + ',{"testKey":"' + test.tags[1].split("@")[1] + '","status":"' + test.state + '", "examples":' + JSON.stringify(test.examples) + ',"comment" : "' + test.comment + '" }';
+                    tests = tests + "" + '{"testKey":"' + test.tags[1].split("@TEST_")[1] + '","status":"' + test.state + '", "examples":' + JSON.stringify(test.examples) + ',"comment" : "' + test.comment + '" }';
+                } else tests = tests + "" + ',{"testKey":"' + test.tags[1].split("@TEST_")[1] + '","status":"' + test.state + '", "examples":' + JSON.stringify(test.examples) + ',"comment" : "' + test.comment + '" }';
```
Comment (medtoure18) [inline index.js]: this is not ok of server version.
the tag, can be '@ZZZ-1122' or something else.
If it's always like '@TEST_' on cloud version, you can also add an if cloud

File: index.js [hunk 12]
```diff
@@ -103,36 +108,96 @@ module.exports = function (config) {
                 if (i == 0) {
                     tests = '"tests" : [';
                     // if this is the first test that is added to the list of tests executed.
-                    tests = tests + "" + '{"testKey":"' + test.tags[2].split("@")[1] + '","status":"' + test.state + '", "examples":' + JSON.stringify(test.examples) + ',"comment" : "' + test.comment + '" }';
-                } else tests = tests + "" + ',{"testKey":"' + test.tags[2].split("@")[1] + '","status":"' + test.state + '", "examples":' + JSON.stringify(test.examples) + ',"comment" : "' + test.comment + '" }';
+                    tests = tests + "" + '{"testKey":"' + test.tags[2].split("@TEST_")[1] + '","status":"' + test.state + '", "examples":' + JSON.stringify(test.examples) + ',"comment" : "' + test.comment + '" }';
```
Comment (medtoure18) [inline index.js]: this is not ok of server version.
the tag, can be '@ZZZ-1122' or something else.
If it's always like '@TEST_' on cloud version, you can also add an if cloud

Comment (medtoure18) [inline index.js]: this is not ok of server version.
the tag, can be '@ZZZ-1122' or something else.
If it's always like '@TEST_' on cloud version, you can also add an if cloud

File: index.js [hunk 13]
```diff
@@ -103,36 +108,96 @@ module.exports = function (config) {
...
+                } else tests = tests + "" + ',{"testKey":"' + test.tags[2].split("@TEST_")[1] + '","status":"' + test.state + '", "examples":' + JSON.stringify(test.examples) + ',"comment" : "' + test.comment + '" }';
```
Comment (medtoure18) [inline index.js]: his is not ok of server version.
the tag, can be '@ZZZ-1122' or something else.
If it's always like '@TEST_' on cloud version, you can also add an if cloud

**Review by medtoure18 — CHANGES_REQUESTED** [review]
@amitasil 
i added some feedbacks. Take a look and get back to me.
beside the "if cloud", you should add more, i think you your work is good.

**Comment by amitasil** [comment]
@medtoure18 

It's too late here :) 

Will do the changes and get back to you tomorrow. 

Thanks!!

**Comment by medtoure18** [comment]
@amitasil 
Ok no problem, have a good night,
 its 2PM here :-)

**Comment by amitasil** [comment]
@medtoure18 

Incorporated your review comments, but the issue still persists :(

**Comment by medtoure18** [comment]
What response body do you get with the second request ! ?

**Comment by amitasil** [comment]
@medtoure18 

Just "object [Object]"

**Comment by medtoure18** [comment]
Try a console.log, it would shows you what's inside the objet

**Comment by amitasil** [comment]
@medtoure18 

Made some updates and it's working fine now!!

This is what I am getting now:

```
SEND TO XRAY => { "testExecutionKey": "SP-10","info" : {"startDate" : "2021-11-25T11:10:10+11:00", "finishDate" :"2021-11-25T11:10:10+11:00","revision": "001","description" : "Results of test execution ", 
"testEnvironments": ["Sandpit"]},"tests" : [{"testKey":"SP-8","status":"PASSED", "examples":[],"comment" : "Successful execution" }]}
XRAY RESPONSE => eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJ0ZW5hbnQiOiJiMmMxZGFhZC0zOGViLTNiY2EtYTFmYy03OWIzODUwOTAwZDEiLCJhY2NvdW50SWQiOiI1NTcwNTg6MjlhMDgyN2YtODY5MC00YWM5LWIzZWUtNDE5MzA5Njg0ZGM4IiwiaXNYZWEiOmZhbHNlLCJpYXQiOjE2Mzc3OTkwMTMsImV4cCI6MTYzNzg4NTQxMywiYXVkIjoiOUI4MjA1QUE0OUMyNDc3MTgwMEM2ODg5NzZBQUZDQ0QiLCJpc3MiOiJjb20ueHBhbmRpdC5wbHVnaW5zLnhyYXkiLCJzdWIiOiI5QjgyMDVBQTQ5QzI0NzcxODAwQzY4ODk3NkFBRkNDRCJ9.wpDfvZ_N5vGBjf5PE9WTKUIYc98qcm52PmA6NHnzpCE
Access token generate successfully!
XRAY RESPONSE => {"id":"10018","key":"SP-10","self":"https://amitrawat.atlassian.net/rest/api/2/issue/10018"}
Test Execution updated successfully!
```

**Review by medtoure18 — APPROVED** [review]
review OK, good job :-)

**Comment by medtoure18** [comment]
@amitasil 
i will create the new npm version this week-end
thank you :-)

**Comment by amitasil** [comment]
Cheers @medtoure18 !! :)

**Comment by medtoure18** [comment]
@amitasil 
just publish the new npm version, can you check if it's ok for you.
i removed request dependancy and updated xray_cloud config as default false.

**Comment by amitasil** [comment]
Thanks @medtoure18 !!

Will try this today.

**Comment by amitasil** [comment]
Hi @medtoure18 , the new changes are looking good.

One observation though. When I am running my two tests in parallel with 'testsExportedFromTestExecution' as false.
I am getting two test runs getting generated in Xray Jira for the two executed tests. Each test run has one test case with the correct executed status.

Can we club all the tests executed in parallel and generate only one test run in Xray Jira, just like a sequential run?

This one could help: https://codecept.io/parallel/#parallel-execution-by-workers

**Comment by medtoure18** [comment]
@amitasil 
Thanks for the test :-)
About parallèle workers, i think we need to change a lot of things. Because we need to know when the last parallèle test is finished before we send the results.  Otherwise we could  creat the test execution empty before all tests.
If you have the time to do another PR you can :-)

**Comment by amitasil** [comment]
Hi @medtoure18 ,

I tried but could find a solution on how to handle Codeceptjs workers :( 

We are planning to run our tests in parallel and upload results back in Xray and this issue has become a blocker for us...

Could you please look into this?

Thanks!!

**Comment by medtoure18** [comment]
Hi @amitasil,
I dont have the time to work on the feature but i can check codeceptjs doc to find where can we find the solution, so it could help you for your PR

**Comment by amitasil** [comment]
Thanks @medtoure18 !!

**Comment by medtoure18** [comment]
Hi @amitasil,

I think the solution is to use the event: **event.all.before** for an empty test execution creation. after that we save the id of  the test execution after each the same execution will be used.
 
right now we are using **event.all.after** to send the results to xray.
see https://codecept.io/hooks/#event-listeners

**Comment by amitasil** [comment]
Hi @medtoure18 

I am trying to use something like this the 'parallel.js' option they have mentioned at https://codecept.io/parallel/#parallel-execution:

But the thing is, I need to run the tests using "node parallel.js" command.

And I am not getting the test tags as an object, see the test details below:

```
{
   "type":"test",
   "title":"Navigate to Business Insurance page @TEST_SP-12 @TESTSET_SP-9 @feature",
   "body":"() => { }",
   "async":0,
   "sync":true,
   "_timeout":2000,
   "_slow":75,
   "_retries":-1,
   "timedOut":false,
   "_currentRetry":0,
   "pending":false,
   "id":"Gp2WtZ5crrM7g8XPpR2h+w",
   "workerIndex":1,
   "retries":-1,
   "duration":5586,
   "err":null,
   "parent":{
      "title":"Navigate to Business Insurance page @REQ_SP-11",
      "ctx":{
         
      },
      "suites":[
         
      ],
      "tests":[
         
      ],
      "root":false,
      "pending":false,
      "_retries":-1,
      "_beforeEach":[
         
      ],
      "_beforeAll":[
         
      ],
      "_afterEach":[
         
      ],
      "_afterAll":[
         
      ],
      "_timeout":2000,
      "_slow":75,
      "_bail":false,
      "_onlyTests":[
         
      ],
      "_onlySuites":[
         
      ],
      "delayed":false,
      "_events":{
         
      },
      "_eventsCount":1
   }
}
```

The above extract comes under this code:

```
workers.on(event.test.after, function (test) {
output.print(JSON.stringify(test));
});
```

Not sure how to get this thing working...

**Pull request merged**